Process monitoring

glasswalk3r opened the pull request:
Simple script to grab information from running process.
Those processes will be filtered by login, and currently a CSV file is generated as output.
Data is still raw (no grouping, no aggregation). I'm still not sure how we want to do that. Some possibilities is to get average and standard deviation, but we should find out if we should group by login and/or process command line (the program that is running).
Collectd has a Perl plugin that could be used to run this script and collect the info, to send to the monitoring host, but this is still to be developed. For now, using crontab would be enough for that, and I'm not sure either we should capture this information long periods.
After all, we are looking for bottlenecks on the whole process. It might be possible we shut it down after some iterations.

Commit:
Initial release

Commit:
Bugs fixes

Added basic testing
Fixed bugs

preaction commented on monitor/Process-Monitor/dist.ini:
```diff
@@ -0,0 +1,22 @@
+name    = Process-Monitor
+author  = Alceu Rodrigues de Freitas Junior <arfreitas@cpan.org>
+license = GPL_3
```
The license for the CPAN Testers project is the terms of Perl 5 itself (Artistic 1 or GPL 1+ at the user's discretion).

glasswalk3r replied:
Done.

preaction commented on monitor/Process-Monitor/dist.ini:
```diff
@@ -0,0 +1,29 @@
+[OurPkgVersion]
+[Test::Kwalitee]
+filename = xt/kwalitee.t
+[ChangelogFromGit]
```
Please do not use this plugin, as it is especially incorrect in this repository which has more than just this distribution inside.

glasswalk3r replied:
Done.

preaction commented on monitor/Process-Monitor/dist.ini:
```diff
@@ -0,0 +1,29 @@
+file_name = Changes
+max_age = 365
+tag_regexp = ^release-(\d+.*)$
+; authordep Dist::Zilla::Plugin::AutoPrereqs
```
You don't need to explicitly declare `authordep` if they're configured (you have an `[AutoPrereqs]` section which declares the dependency).

glasswalk3r replied:
Done.

preaction commented on monitor/Process-Monitor/lib/Process/Monitor.pm:
```diff
@@ -0,0 +1,150 @@
+use strict;
+use warnings;
+use Proc::ProcessTable 0.53;
+use List::MoreUtils 0.416 'any';
```
Please use the `any` that is shipped with List::Util v1.33. We use Perl 5.24, so that version of List::Util is in core. The List::MoreUtils distribution is becoming harder to install.

glasswalk3r replied:
Done.

preaction reviewed (changes requested):
This looks great so far. Thanks for writing documentation and tests.

I think with how much code there is that it would be easier to deploy if this was just a single script without an accompanying library. I don't think that function in the library will get much use outside the script that's using it, and the installation toolchain for the module makes things more complicated than I think they need to be.

Even if it's a single script, you can still have a module to test by making the script have a `package` line. Here's an example of a "runnable module" or "modulino": https://github.com/ChicagoPM/Command-Line-Apps/blob/master/modulino_template.pm. The line `exit main( \@ARGV ) unless caller` is what makes the module loadable without running the script inside (the script is the `main` subroutine). That repository also has an example of how to load/test a script like that (greet.pm and greet.t).

Commit:
Requested changes for Merge pull request

Replaced List::MoreUtils with List::Utils, which is core to more recent
perl versions
Removed unecessary authordep Dist::Zilla::Plugin::AutoPrereqs
Removed Dist::Zilla plugin ChangelogFromGit since it would take any
changes from the repository
Changed code license from GPL3 to GPL1

glasswalk3r commented:
### Considerations about using a modulino or a single script

- I could have written a single script (and I did it, it is on cpantesters home, it is called "sample.pl"), but then I would have to install dependencies that CPANFile will take care of (and also the other nice features like unit testing).
- A modulino is a nice idea, but I'm not sure of the benefits: the package itself has a `sub` to separate the information retrieval from the usage itself (now it uses CSV as output, but could be anything else in the future, or having multiple choices to be selected as a parameter). The associated script could do it all by itself (but then I lose the benefits of having a distribution) or I could write it as a modulino to avoid having the script at all, but if I got it correctly, I would have to type something like:

```
perl -MProcess::Monitor --user=barbie --user=www-data
```

Instead of:

```
procmon --user=barbie --user=www-data
```

Right now the users are hard coded in the `procmon` script (ugly, I know), but I could setup then as default values if there are no options given. Considering that implemented as a modulino, I would have to give another option as `--default` to have them applied as the users instead of specifying them with `--user`.

What do you think?

preaction commented:
The modulino is a runnable script. The example I linked you to has a shebang line (`#!/usr/bin/env perl`), so it can be executed directly: `./script.pm`. Command-line arguments can be given (`./script.pm --foo bar baz`), and the example I linked to shows how to parse command-line options. The script is still testable, like I said. Here's an article by bdfoy explaining in more detail: http://perltricks.com/article/107/2014/8/7/Rescue-legacy-code-with-modulinos/

The benefits are that the code is more easily deployable: A single file is copied into the `~/bin` directory, and then `cpanm <dependencies>` is run (these two things are put in the `Rexfile`). I'm not sure how making it a single script makes it more difficult to add output formats, or even why that is a thing the CPAN Testers project will need.

If you want to make it a distribution, make it a distribution outside of the CPAN Testers project. I don't know what the benefits of making this into a distribution are, though.

> Considering that implemented as a modulino, I would have to give another option as --default to have them applied as the users instead of specifying them with --user

I also do not understand this. The modulino is able to parse command-line arguments exactly like any other script.

I'm starting to wonder: What will this monitor will do that cannot be done any other way?

glasswalk3r commented:
> If you want to make it a distribution, make it a distribution outside of the CPAN Testers project. I don't know what the benefits of making this into a distribution are, though.

I don't need to make another distribution, I just thought initially that separating data gathering from the real application of it would be interesting. **And that's OK if the design decision is to go with a Modulino**, never passed to my mind that discussing implementation details was an offensive or useless and time consuming activity.

> I'm starting to wonder: What will this monitor will do that cannot be done any other way?

I understood that currently we have absolute nothing to monitor processes from a specific(s) user(s), if we want finner grained details (and I think we would want, if we are interested in finding out that are the processes consuming more CPU, understanding better how they relate to each other and being able to measure any performance benefit after changes were implemented).

Implementation can be made in different ways, after all this is Linux and Perl. I'm willing to implement a Modulino, but if you're not sure if this kind of monitoring is really desired or not, that's up to you.
